hardware-health.md

```diff
@@ -0,0 +1,307 @@
+    nm{{NATS message bus}}
+    nkv[(NATS KV)]
+ end
+
+ subgraph Spectral
+    agentRollup[\Rollup-Agent/]
+  end
+
+  agentRollup -- Watch & rollup --> nkv
+  agentBMC -- run checks, collect metrics --> bmc1
+  agentBMC -- write results --> nm
+```
+
+The health roll-up agent watches Server health information as it is updated by an Agent and updates the Server health roll-up information.
+
+TODO: elaborate a bit more as the concept is hashed out.
+
+
+### Check plugins
+---
+
+Checks are to be implemented as [plugins](https://github.com/hashicorp/go-plugin/tree/main) by the operator,
+these plugins are grouped by the kind of agent and are made available to the agents for execution.
+
+Various checks that could be implemented are listed below,
+
+```mermaid
+classDiagram
+namespace BMCAgent {
+  class BMC-Checks {
+    - HTTPs, SSH Available
+    - Credentials function
+    - Redfish Health data
+    - Export Power metrics
+    - Export Thermal metrics
+    - Export Fan speeds
+    - Event log errors
+    - Firmware updated
+    - BIOS configured
+    - ...
+  }
+}
+
+namespace OSAgent {
+  class Server-Checks {
+    - CPU exceptions
+    - Memory exceptions
+    - Disk SMART errors
+    - NIC speeds, Errors
+    - Firmware updated
+    - BIOS configured
+    - ...
+  }
+}
+
+namespace CustomAgent {
+  class ServiceDependency-Checks {
+    - DHCP available
+    - Switch Port errors
+    - FC light levels
+    - Switch ARP
+    - Cabling
+    - Provisioning services available
+    - ...
+  }
+}
+
+namespace PDUAgent {
+    class PDUChecks {
+     - Export Power metrics
+    }
+}
+```
+
+
+### Check definitions
+---
+
+The spec for check definition is to be formalized, these are just example definitions,
+the definition is the runtime configuration for Spectral agents and is defined by operators.
+
+Sample check definitions for the `BMCAgent`
+```yaml
+defaults:
+    kind: "bmcAgent"
+    # https://godoc.org/github.com/robfig/cron#hdr-CRON_Expression_Format
+    schedule: "0 5 */4 * * *"
+
+checks:
+  - bmc_network:
+      opdoc: "Check if the BMC network port is up and has no errors"
+      owner: ["dcops"]
+      tags: ["bmc", "network"]
+
+   - bmc_available:
+      opdoc: "BMC IP Address reachability and credential check"
+      owner: ["dcops"]
+      tags: ["bmc","network"]
+      depends_on: ["bmc_network"]
+
+   - bios_configured:
+     opdoc: "Check the BIOS configuration is as expected"
+     owner: ["dcops"]
+     tags: ["bmc","network"]
+     depends_on: ["bmc_network"]
+
+   - firmware_updated:
+     opdoc: "Check the firmware installed is up to date"
+     owner: ["fleet"]
+     tags: ["bmc","firmware"]
+     depends_on: ["bmc_network"]
+```
+
+
+
+## References
+
+#### Existing server hardware health monitoring systems
+ - [Google's Hardware health monitoring system - see Chapter 7. Dealing with Failures and Repairs in Datcenter as a computer - PDF](https://www.morganclaypool.com/doi/10.2200/S00874ED3V01Y201809CAC046)
```

Comment:
This link is not working for me.

Reply:
Dustin pointed out a working link for the reference: https://pages.cs.wisc.edu/~shivaram/cs744-readings/dc-computer-v3.pdf

Reply:
Thanks! pushed a fix to update the link